File: src/main/java/rag/answer/FallbackHandler.java
package rag.answer;

import rag.app.Context;
import rag.intents.Intent;

import java.util.List;

public class FallbackHandler {

    public Answer buildFallback(Context context) {
        String reason = determineReason(context);
        context.setFallbackReason(reason);
        String message = switch (reason) {
            case "EmptyQuestion" -> "I couldn't produce an answer because the question was empty.";
            case "UnknownIntent" -> "I am unsure which topic you are asking about. Please rephrase the question.";
            default -> "I could not produce an answer for this query.";
        };
        return new Answer(message, List.of("fallback:" + reason));
    }

    private String determineReason(Context context) {
        if (context.getQuestion() == null || context.getQuestion().isBlank()) {
            return "EmptyQuestion";
        }
        Intent intent = context.getIntent();
        if (intent == null || intent == Intent.Unknown) {
            return "UnknownIntent";
        }

        return "Unknown";
    }
}

File: src/main/python/rag/answer/fallback_handler.py
from typing import List, Optional
from rag.answer.answer import Answer
from rag.app.context import Context
from rag.intents.intent import Intent


class FallbackHandler:
    def build_fallback(self, context: Context) -> Answer:
        reason: str = self._determine_reason(context)
        context.set_fallback_reason(reason)
        message: str = self._get_message(reason)
        return Answer(message, [f"fallback:{reason}"])

    def _determine_reason(self, context: Context) -> str:
        if not context.get_question() or not context.get_question().strip():
            return "EmptyQuestion"
        intent: Optional[Intent] = context.get_intent()
        if intent is None or intent == Intent.Unknown:
            return "UnknownIntent"
        return "Unknown"

    def _get_message(self, reason: str) -> str:
        if reason == "EmptyQuestion":
            return "I couldn't produce an answer because the question was empty."
        elif reason == "UnknownIntent":
            return "I am unsure which topic you are asking about. Please rephrase the question."
        else:
            return "I could not produce an answer for this query."

